### ISSUE TITLE
[libc++][std::allocator]

### ISSUE THREAD

**Issue by sivan-shani** [llvm/llvm-project#113568]
According to ISO/IEC 14882:2014 under clause 17.6.3.5, 'Allocator requirements', Table 28 — 'Allocator requirements', one of the Allocator requirements is 'rebind'.

While 'libcxx/include/__memory/allocator.h' implement a rebind for the specialized void allocator, it does not implemented for the general case.

This has the effect when a custom allocator is being derived from std::allocator. 'rebind' has to be implemented in the derived class, as can be seen in this example: [godbolt](https://godbolt.org/z/rq7h8ozbY)

The first instance of a derived custom allocator works, because it is implementing 'rebind', while the second fail.

**Comment by ldionne** [comment]
I believe this behaves correctly. `std::allocator<T>` does provide a `rebind` member in C++14, and that member is such that:

```
std::allocator<T>::rebind<U>::other == std::allocator<U>
```

For example: https://godbolt.org/z/rvffW4583

That's what the Standard specifies, and that's the only thing it can do. It seems that what you would want is for the `rebind` member in the base class to automatically rebind the derived class, but that's both not what the Standard says and also not implementable at all with the provided API.

Both libc++ and libstdc++ behave the same, which makes me think this analysis is correct: https://godbolt.org/z/89hE6MYPs

**Comment by philnik777** [comment]
Yes, this is 100% expected behaviour. C++20 removed the `rebind` member in `std::allocator`, which solves the problem for derived allocators.

**Comment by sivan-shani** [comment]
Thank you for your attention. TBH, I can not 100% decipher the standard meaning, but your interpretation seems reasonable. 
The reason that I assumed that rebind should be inheritable is seeing the usage of std::allocator done by a reputable c++ expert body, which was assuming that rebind will be inherited by the derived class, but afterward I saw that in newer versions of their product they actually implementing it in the derived class.
Many thanks for the reply.

**Comment by ldionne** [comment]
Just to clarify, the `rebind` member itself will be inherited by the derived class, however it's definition is going to be incorrect since it will rebind the base class type, not the derived class type. The error you were seeing wasn't caused by `rebind` being non-existent, it was caused by it not having the right value.
